File: packages/adapter-android/app/src/test/kotlin/com/thrw/adapter/android/triggers/RecordingEventLifecycle.kt
package com.thrw.adapter.android.triggers

import com.thrw.adapter.android.protocol.EventKind
import com.thrw.adapter.android.protocol.Priority

/** One reported trigger transition, in order, for the monitor tests. */
sealed interface TriggerCall {
    val kind: EventKind

    data class Started(override val kind: EventKind, val priority: Priority) : TriggerCall

    data class Ended(override val kind: EventKind) : TriggerCall
}

/**
 * Records what a monitor reported to its node. Shared by
 * `CallTriggerMonitorTest` and `VoipTriggerMonitorTest`; that the calls
 * actually reach MQTT is `AndroidNodeTest`'s job.
 */
class RecordingEventLifecycle : EventLifecycle {
    val calls = mutableListOf<TriggerCall>()

    /**
     * #234. Mirrors `AndroidNode`'s own `activeEvents` closely enough
     * for `ManualClaim` to be tested against it: a trigger is active
     * from a successful `emitEvent` until an `endEvent`. A
     * `LinkedHashSet` for the same reason the node uses one - insertion
     * order is what `mostRecentTrigger()` reads.
     */
    private val active = linkedSetOf<EventKind>()

    override suspend fun emitEvent(type: EventKind, priority: Priority) {
        calls += TriggerCall.Started(type, priority)
        active += type
    }

    override suspend fun endEvent(type: EventKind) {
        calls += TriggerCall.Ended(type)
        active -= type
    }

    override fun isEventActive(type: EventKind): Boolean = active.contains(type)

    override fun activeEventKinds(): List<EventKind> = active.toList()
}

File: packages/adapter-mac/Tests/AdapterMacTests/Triggers/RecordingEventLifecycle.swift
import Foundation

@testable import AdapterMac

/// Records every `emitEvent`/`endEvent` call instead of publishing
/// anything real - mirrors `adapter-android`'s
/// `RecordingEventLifecycle.kt` fixture.
final class RecordingEventLifecycle: EventLifecycle, @unchecked Sendable {
    struct Emitted: Equatable {
        let type: EventKind
        let priority: Priority
    }

    private(set) var emitted: [Emitted] = []
    private(set) var ended: [EventKind] = []

    /// #234. Mirrors `MacNode`'s own `activeEvents` closely enough for
    /// ``ManualClaim`` to be tested against it: a trigger is active from
    /// a successful `emitEvent` until an `endEvent`. Insertion order is
    /// preserved for the same reason `ActiveEventSet` preserves it.
    private var active: [EventKind] = []

    func emitEvent(type: EventKind, priority: Priority) async throws {
        emitted.append(Emitted(type: type, priority: priority))
        if !active.contains(type) { active.append(type) }
    }

    func endEvent(type: EventKind) async throws {
        ended.append(type)
        active.removeAll { $0 == type }
    }

    func isEventActive(_ type: EventKind) -> Bool {
        active.contains(type)
    }

    func activeEventKinds() -> [EventKind] {
        active
    }
}
